### PR TITLE
[RDY] OS X: Package.cpath and luarocks

### PR DESCRIPTION
**Pull request by tobylane**
First commit adds the relevant folders of a local luarocks installation to the package.cpath var. The working directory of a process launched as an app is $HOME so no relative path will ever work.

Second commit adds a cmake option to turn on a post build rule, another always on post build rule and some comments edits. The first includes dylibs (dependancies) by copying them in and linking CorsixTH  to them and each other. It's assumed the user installed them by homebrew, others can be added when users emerge. The second script creates a local luarocks installation and patches Luasocket.

Output of osx_dylibs https://travis-ci.org/tobylane/CorsixTH/jobs/119871660#L1255 and osx_luarocks https://travis-ci.org/tobylane/CorsixTH/jobs/119871660#L978

### CODE REVIEW

**Comment by TheCycoONE** [comment]
Maybe make your second commit a make target (like docs) instead of a cmake flag? Or would that not work?

**Comment by tobylane** [comment]
I believe it would have to be a duplicated CorsixTH sub project definition as CMake projects can't be cloned/extended, or the user would run CMake a second time after Xcode.

**Comment by tobylane** [comment]
Added a third commit that does the dylib change for AnimViewer. Output https://travis-ci.org/tobylane/CorsixTH/jobs/120023001#L1488

**Comment by TheCycoONE** [comment]
I meant `ADD_CUSTOM_TARGET`, or better yet `INSTAL`' and then after building `make foo` or `make install`.

All of the rsync steps should be in the install target. 

I believe `INSTALL` can already do most of the dylib  changes for you too.

For xcode I have seen it run as `xcodebuild --target install`

**Comment by tobylane** [comment]
Changes per request. Install as a target is already taken and can't be reliably reclaimed so they are named IncludeDylibs and IncludeAnimViewDylibs. They can be chained in a command with the app build like `xcodebuild -configuration Release -target CorsixTH -target IncludeDylibs`

**Comment by TheCycoONE** [comment]
Ok, I'm fine with this, or maybe just a ```redistributable``` target, as includeDylibs is a bit wordy and specific.

File: CorsixTH/CorsixTH.lua [hunk 1]
```diff
@@ -17,6 +17,14 @@ for _, arg in ipairs({...}) do
   end
 end
 
+-- On Mac OS X add the local luarocks folder to package.cpath.
+if (io.popen"uname":read'*l' == "Darwin") then
```
Comment (TheCycoONE) [inline CorsixTH/CorsixTH.lua]: Should use brackets for popen and read. Could potentially be used everywhere not just OSX, so maybe this outer condition isn't necessary.

File: scripts/osx_luarocks [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+elif [ "$lua" = "" ]; then lua=5.2
+fi
+
+if [ -d luarocks/share/lua/"$lua" ]
+  then echo "Luarocks for Lua"$lua" in place."
+  exit
+elif [ -d "$dir/share/lua/$lua/socket/" ] && [ -z "$1" ]
+  then echo "Copying luarocks from global installation."
+  mkdir -p luarocks/share/lua
+  cp -R "$dir/share/lua/$lua" luarocks/share/lua/
```
Comment (maki-chan) [inline scripts/osx_luarocks]: Missing copy command for libraries here, should also have something like:

`cp -R "$dir/lib/lua/$lua" luarocks/lib/lua/`

as the path `luarocks/lib/lua` was defined for the various needed Lua packages in the `package.cpath` variable.

**Comment by tobylane** [comment]
@TheCycoONE: Will do.

@maki-chan: I started off doing that but found it wasn't necessary. I renamed the share folder in your build, removed my local copy of luafilesystem and the game works fine. We can investigate this further, would you like to use IRC? We're at FreeNode #corsix-th.

**Comment by TheCycoONE** [comment]
If lfs.so doesn't exist in your lua path there is no way CorsixTH would work, so it must be somewhere on your system.

Actually I was thinking about it, and those lua so files may have been what was giving the bad instruction errors.

**Comment by clobber** [comment]
@tobylane Still working on getting this merged? This looked promising.

File: CorsixTH/CorsixTH.lua [hunk 3]
```diff
@@ -17,6 +17,12 @@ for _, arg in ipairs({...}) do
   end
 end
 
+-- Add the local luarocks folder to package.cpath.
+local luarocks_path = debug.getinfo(1, "S").source:sub(2):match("(.*/)") ..
+    "luarocks/lib/lua/" .. string.sub(_VERSION,5)
+package.cpath = luarocks_path .. "/?.so;" .. luarocks_path .. "/mime/?.so;" ..
```
Comment (TheCycoONE) [inline CorsixTH/CorsixTH.lua]: A little late now, but /mime/?.so and socket/?.so shouldn't have been part of this. ? is replaced by the module name, which is mime.core and socket.core.

**Comment by tobylane** [comment]
Redone to be a part of the install process, simplified the lua part.

**Comment by TheCycoONE** [comment]
Sorry @tobylane I didn't realise that you had updated this. It needs a CMAKE flag I think - we don't want to unconditionally brew install anything. I'll review it closer soon.

**Comment by TheCycoONE** [comment]
I still feel this needs a toggle in cmake. Everywhere else we download external tools we opt in e.g. `USE_VCPKG`

**Comment by tobylane** [comment]
Fixed, and removed the reference to homebrew.

File: CorsixTH/CMakeLists.txt [hunk 4]
```diff
@@ -279,5 +279,10 @@ if(NOT USE_SOURCE_DATADIRS)
       SET(BU_CHMOD_BUNDLE_ITEMS ON)
       FIXUP_BUNDLE(${CMAKE_INSTALL_PREFIX}/CorsixTH.app \"\" \"\")
       ")
+    if(WITH_LUAROCKS)
+      install(CODE "execute_process(
+        COMMAND bash -c ${CMAKE_SOURCE_DIR}/scripts/macos_luarocks ${CMAKE_INSTALL_PREFIX})
```
Comment (colinjmatt) [inline CorsixTH/CMakeLists.txt:284]: In order for bash to handle spaces in the file path, this would be better as
`COMMAND bash \"${CMAKE_SOURCE_DIR}/scripts/macos_luarocks\" \"${CMAKE_INSTALL_PREFIX}\")`

**Comment by tobylane** [comment]
Fixed, thanks.

Ready.

File: CorsixTH/CMakeLists.txt [hunk 5]
```diff
@@ -277,7 +277,12 @@ if(NOT USE_SOURCE_DATADIRS)
     install(CODE "
       INCLUDE(BundleUtilities)
       SET(BU_CHMOD_BUNDLE_ITEMS ON)
-      FIXUP_BUNDLE(${CMAKE_INSTALL_PREFIX}/CorsixTH.app \"\" \"\")
+      FIXUP_BUNDLE(\"${CMAKE_INSTALL_PREFIX}/CorsixTH.app\" \"\" \"\")
```
Comment (colinjmatt) [inline CorsixTH/CMakeLists.txt:280]: Is adding quotes here needed as that's CMake code and should be handled fine? They were needed on line 284 because bash doesn't handle unquoted filepaths with spaces.

Someone who knows more abotu CMake would need to comment on that.

Comment (tobylane) [reply CorsixTH/CMakeLists.txt:280]: > -- Installing: /Applications/test path/CorsixTH.app/Contents/Resources/LICENSE.txt
> -- fixup_bundle
> --   app='/Applications/test'
> --   libs='path/CorsixTH.app'
> --   dirs=''
> --   ignoreItems=''
> -- warning: *NOT* handled - directory/file does not exist...
> CMake Error at /usr/local/Cellar/cmake/3.17.3/share/cmake/Modules/BundleUtilities.cmake:988 (message):
>   error: fixup_bundle: not a valid bundle
> Call Stack (most recent call first):
>   cmake_install.cmake:62 (FIXUP_BUNDLE)
> -- fixup_bundle: done

Without that, I get this error where CMake has the same problem.

Comment (colinjmatt) [reply CorsixTH/CMakeLists.txt:280]: Ah yes that makes sense. It might be worthwhile as a separate PR to go through and quote all filepath variables to prevent this anywhere else in CMake.
